README.md
![Head JS](http://headjs.com/site/assets/img/logo-big.png)

This project was never announced. `git push` and it was all viral.

[![Build Status](https://travis-ci.org/headjs/headjs.png?branch=master)](https://travis-ci.org/headjs/headjs)

##About
  * _Load scripts and css on demand_
  * _Achieve responsive design with CSS that targets different screen resolutions, paths, states and browsers_
  * _Detect various browsers and their features_
  * _Target HTML5 and CSS3 safely_
  * __Make it the only script in your HEAD__
  * __A concise solution to universal problems__


##Where
* __WebSite__
  * [http://headjs.com](http://headjs.com)
* __Blog/Dev Updates__
  * [http://www.i-technology.net](http://www.i-technology.net/search/label/headjs)
* __Bugs/Support__
  * [https://github.com/headjs/headjs/issues](https://github.com/headjs/headjs/issues)
* __Get Community Help__
  * [http://stackoverflow.com](http://stackoverflow.com/questions/tagged/head.js)
  * _Remember to tag your questions with_: head.js
* __Vote for Features/Suggestions__
  * [http://headjs.uservoice.com](http://headjs.uservoice.com/forums/182905-headjs-feature-suggestions)

##Updates

###0.99 / 2012-11-15

	- Load: Fixed regression in IE6, caused by IE10 fix
	- Load: CSS loading seems to work in all browsers.
		- However a few will not trigger the callback. Over 90% do.
		- For now only load css in situations where you don't need the callback triggered.
		- head.load("file1.css", "file2.css")
	- Load: Conditional loading with head.test() now in evaluation phase
		- try it, but don't rely on it yet
		- head.test(bool, "ok.js", "failed.js", callback)
	- All: CDN is now availiable thanks to: http://cloudflare.com
		- Info in download section on main site
	- Unit Tests
		- Integrated with main site so that everyone can participate
		- They have also been hooked up to automatically report stats back to http://browserscope.org


###v0.98 / 2012-11-09

	- Load: Fixed loading bug in IE10
	- Load: Corrected some issues with loading from inside <head>
	- Load: Rewrite of large parts of code base
	  - Started to massively document the sourcecode :)
	- Load: Test Only (css loading)
	  - You can now load CSS files with head.load("somefile.css")
	  - Consider this as under testing
	  - head.load is an alias to head.js
	- Css3: moved "touch" detection from core to here
	- Css3: added "retina" detection
	- Css3: replaced "font-face" detection that was using "Conditional Comments" with simplisitc browser version detection
	- Core: Added gt, gte, lte, eq classes to width detection (lt existed already)
	- Core: Added gt, gte, lt, lte, eq classes for browser vendor & version detection
	- By default only lt/gt classes are activated
	  - You can of course configure to your likings via head_conf


###RoadMap (v1.0 / v2.0) 2012-11-06

	- Conditional Loader
	  - head.test(bool, success, failure, callback)
	- Resource Loading
	  - head.load(somefile.js , callback)
	  - head.load(somefile.css, callback)
	- Advanced Responsive Design
	  - Take into account browsers min/max versions & height/width pseudo media queries (lt, lte, eq, gt, gte)
	  - .ie-lt6, .ff-gte17, .w-lt1024, .h-gte800
	- Inverse naming conventions for readability & to be able to programatically inject values more easily
	  - .no-box-shadow -> .box-shadow-true/false
		- jQuery(selector).addClass("box-shadow-" + head.features.boxShadow)
		- class="localStorage-<%= bool.toString().toLower() %>"
	  - .no-ie -> .ie-true/false
	  - .index-page -> .page-index
	- Advanced Configuration  
	  - Expose a maximum of functionalty through head_conf, so you only get the functionality you actually require
	    - Enable/Disable lt/gt class generation
	    - Enable/Disable height/width support
   	    - Configure custom height/width/browser breakpoints 
	    - ..etc
	- Remove tests from css3.js and let users "roll their own"
      - 80% of css related tests are probably irrevelant anyways
        - For things like border-radius & the likes everyone uses gracefull degradation with their css anyways ..right ? :)
	- Renaming files
	  - head.responsive.js (can be used standalone)
	    - old core.js with a few additions from css3.js so we can externalise the test suite
	  - head.load.js (can be used standalone)
	    - old load.js
	  - head.features.js (MUST be used with head.responsive.js)
	    - old css3.js, but now only integrates actual tests	  
	- Revamped website with Twitter Bootstrap
	  - Make the site a tad more ergonomic to navigate
	  - Expose unit tests & advanced documentation directly in site
	  - Custom "roll-your-own" builder
	  - Integrate website with gh-pages for quicker deploy
	  
*How much of this roadmap will end up in the current version of head.js, i don't know yet.*

* Most likely when v1.0 comes out, it will have a good part of the above functioanlity included while at the same time preserving the existing api syntax.
  * Like this, those that already depend on HeadJS get no breaking changes.
* At the same time there will probably be a v2.0 release that uses the above syntax, and be the defacto version for future development.
	  
__P.S. 70% of the preliminary work mentioned above is already done :)__

###v0.97a / 2012-10-20

	- Updated QUnit &amp; got unit tests running again
	- Swictched to "use strict"
	- Fixed up some variable usages
	- Added browser detections other than just for ie-lt
	- updated browser regexes (firefox, safari, opera, ios, android, webkit)
	- detect if browser is: desktop, mobile, touch enabled
	- detect portrait/landscape mode
	- html5 shim now only triggers on ie-lt9
	- added a throttle to onResize, since some browsers fire tons of events/sec
	- added corrected height/width measurements, but only exposed via new object: head.screen
	  - contains height/width, innerHeight/innerWidth, outerHeight/outerWidth
	- forced all css router names to lowercase just in case ppl type in url's with wierd casings

changelog.txt
1.0 :: Items marked with (**) are considered "safe" changes and have been silently integrated into /src & /dist in v0.99 on GitHub ..and maybe on CDN
 - Improved CSS Router (**)
   - https://github.com/headjs/headjs/issues/227
 - Added "main" HTML5 element to shim (**)
   - https://github.com/headjs/headjs/pull/230
 - Always add small shortcut to browser & version no matter what browser breakpoints configured (**)
   - Makes it possible to work without any breakpoints at all
 - Load file via data attribute on HeadJS script tag (**)
   - data-headjs-load="...filetoload.ext"
   - https://github.com/headjs/headjs/pull/213
 - Add options as a property label
   - https://github.com/headjs/headjs/pull/215
 - Allow ASYNC loading
   - https://github.com/headjs/headjs/issues/204
 - Prevent loading empty strings (**)
   - https://github.com/headjs/headjs/pull/184
 - Fixed CSS classes getting bigger on successive resizes under Chrome (**)
   - https://github.com/headjs/headjs/issues/226
 - Cleaned up a bunch of syntaxt to conform to JSHint (**)
   - now using a very strict .jshintrc
 - Added missing .gitattributes (**)
 - Fixed "Invalid regular Expression (**)
   - https://github.com/headjs/headjs/issues/255
- Changed window.frameElement detection (**)
  - https://github.com/headjs/headjs/pull/257
- Source map files have been added for all minified JS files (**)
- ..in progress

0.99
 - Load: Fixed regression in IE6, caused by IE10 fix
 - Load: CSS loading seems to work in all browsers.
   - However a few will not trigger the callback. Over 90% do.
   - Either don't use it, or only load css in situations when you don't need the callback triggered.
 - Load: Conditional loading with head.test() now in evaluation phase
   - try it, but don't rely on it yet
   - head.test(bool, "ok.js", "failed.js", callback)
 - All: CDN is now availiable thanks to: http://cloudflare.com
   - Info in download section on main site
 - Unit Tests
   - Integrated with main site so that everyone can participate
   - They have also been hooked up to automatically report stats back to http://browserscope.org

0.98
- Load: Fixed loading bug in IE10
- Load: Corrected some issues with loading from inside &lt;head&gt;
- Load: Rewrite of large parts of code base
  - Started to massively document the sourcecode :)
- Css3: moved "touch" detection from core to here
- Css3: added "retina" detection
- Css3: replaced "font-face" detection that was using "Conditional Comments" with simplisitc browser version detection
- Core: Added gt, gte, lte, eq classes to width detection (lt existed already)
- Core: Added gt, gte, lt, lte, eq classes for browser vendor & version detection
- By default only lt/gt classes are activated
  - You can of course configure to your likings via head_conf

0.97a
- Updated QUnit &amp; got unit tests running again
- Swictched to "use strict"
- Fixed up some variable usage
- Added browser detections other than just for ie-lt
- updated browser regexes (firefox, safari, opera, ios, android, webkit)
- detect if browser is: desktop, mobile, touch enabled
- detect portrait/landscape mode
- html5 shim now only triggers on ie-lt9
- added a throttle to onResize, since some browsers fire tons of events/sec
- added corrected height/width measurements, but only exposed via new object: head.screen
  - contains height/width, innerHeight/innerWidth, outerHeight/outerWidth
- force all css router names to lowercase just in case ppl try typing in names with wierd casings

test/unit/test.load.js
module('Load');

// INFO: Never link to raw.github.com since that returns text/plain instead of application-x/javascript headers ..which breaks IE
// INFO: Do not link 2 same libraries twice, or it will already be in cache
// So now for a big ass helper so we can load some libs more than once ..anyways !
var libs = (function(window, undefined) {
    var counter = {};

    function count(item) {
        if (counter[item] === undefined) {
          counter[item] = 0;
        }

        return counter[item]++;
    }

    function mootools() {
        window.$$ = undefined;
        return 'http://ajax.googleapis.com/ajax/libs/mootools/1.4.5/mootools-yui-compressed.js?' + count("mootools");
    }
    function jquery() {
        window.$      = undefined;
        window.jQuery = undefined;
        return 'http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js?' + count("jquery");
    }
    function jshint() {
        window.JSHINT = undefined;
        return 'http://ajax.aspnetcdn.com/ajax/jshint/r07/jshint.js?' + count("jshint");
    }
    function knockout() {
        window.ko = undefined;
        return 'http://ajax.aspnetcdn.com/ajax/knockout/knockout-2.2.1.js?' + count("knockout");
    }
    function spin() {
        window.Spinner = undefined;
        return 'http://fgnass.github.io/spin.js/dist/spin.min.js?' + count("spin");
    }
    function stapes() {
        window.Stapes = undefined;
        return 'http://hay.github.io/stapes/stapes.min.js?' + count("stapes");
    }
    function notificon() {
        window.Notificon = undefined;
        return 'http://makeable.github.io/Notificon/notificon.js?' + count("notificon");
    }
    function sly() {
        // depends on jQuery
        window.Sly = undefined;
        return 'http://darsa.in/sly/js/sly.min.js?' + count("sly");
    }
    function underscore() {
        window._ = undefined;
        return 'http://underscorejs.org/underscore-min.js?' + count("underscore");
    }

    // Exports
    return {
        mootools  : mootools,
        jquery    : jquery,
        jshint    : jshint,
        knockout  : knockout,
        spin      : spin,
        stapes    : stapes,
        notificon : notificon,
        sly       : sly,
        underscore: underscore
    };
})(window);

asyncTest("load(jsFilePath, jsFilePath, callback)", function() {
    expect(2);

    head.js(
        libs.mootools(),
        libs.jquery(),


        function() {
            $j = jQuery.noConflict();

            ok(!!$j("#qunit-header").addClass, "Loaded: jQuery");
            ok(!!$$("#qunit-header").addClass, "Loaded: Mootools");

            start();
        }
    );
});

asyncTest("ready(jsFileName).load(jsFilePath)", function () {
    expect(1);

    head.ready("jquery.min.js", function() {
        ok(!!jQuery, "Ready: jquery.min.js");

        start();
    })

    .load(libs.jquery());
});


asyncTest("load({ label: jsFilePath }, { label: jsFilePath }, callback).ready(label, callback)", function (assert) {
    expect(6);

    head.load(
        { jshint  : libs.jshint() },
        { jquery  : libs.jquery() },
        { knockout: libs.knockout() }
    )

    .ready("jshint", function () {
        ok(!!JSHINT, "Ready: jshint");
        assert.step(1, "Step 1: jshint");
    })

    .ready("jquery", function () {
        ok(!!jQuery, "Ready: jquery");
        assert.step(2, "Step 2: jquery");
    })

    .ready("knockout", function () {
        ok(!!ko, "Ready: knockout");
        assert.step(3, "Step 3: knockout");

        start();
    });
});


test/unit/test.feature.js
/**
 * Test Feature Detection
 */
module('Features');

var classes = " " + document.documentElement.className;

function hasClass(css) {
    return classes.indexOf(" " + css) !== -1;
}

test("hasClass(className)", function () {
    expect(1);

    ok(hasClass('boxshadow') || hasClass('no-boxshadow'), 'box-shadow');
});

test/unit/test.load-css.js
module('Load');

function getStyle(ele, styleProp) {
    var y = "";

    if (ele.currentStyle) {
        y = ele.currentStyle[styleProp];
    }
    else if (window.getComputedStyle) {
        y = document.defaultView.getComputedStyle(ele, null).getPropertyValue(styleProp);
    }

    return y;
}

// INFO: will make had fail (and nothing else continues!) if file not exists
asyncTest("ready(cssFileName).load(cssFilePath)", function () {
    expect(1);

    head.ready("test.css", function () {
        var result = getStyle(document.getElementById("browserscope"), "display");
        ok(result === "block", "Ready: test.css");

        start();
    })

    .load("assets/test.css");
});


src/load.js
/*!
 * HeadJS     The only script in your <HEAD>
 * Author     Tero Piirainen  (tipiirai)
 * Maintainer Robert Hoffmann (itechnology)
 * License    MIT / http://bit.ly/mit-license
 *
 * Version 0.99
 * http://headjs.com
 */
; (function (win, undefined) {
    "use strict";

    var doc = win.document,
        domWaiters = [],
        queue      = [], // waiters for the "head ready" event
        handlers   = {}, // user functions waiting for events
        assets     = {}, // loadable items in various states
        isAsync    = "async" in doc.createElement("script") || "MozAppearance" in doc.documentElement.style || win.opera,
        isHeadReady,
        isDomReady,

        /*** public API ***/
        headVar = win.head_conf && win.head_conf.head || "head",
        api     = win[headVar] = (win[headVar] || function () { api.ready.apply(null, arguments); }),

        // states
        PRELOADING = 1,
        PRELOADED  = 2,
        LOADING    = 3,
        LOADED     = 4;

    //#region PRIVATE functions

    //#region Helper functions
    function noop() {
        // does nothing
    }

    function each(arr, callback) {
        if (!arr) {
            return;
        }

        // arguments special type
        if (typeof arr === "object") {
            arr = [].slice.call(arr);
        }

        // do the job
        for (var i = 0, l = arr.length; i < l; i++) {
            callback.call(arr, arr[i], i);
        }
    }

    /* A must read: http://bonsaiden.github.com/JavaScript-Garden
     ************************************************************/
    function is(type, obj) {
        var clas = Object.prototype.toString.call(obj).slice(8, -1);
        return obj !== undefined && obj !== null && clas === type;
    }

    function isFunction(item) {
        return is("Function", item);
    }

    function isArray(item) {
        return is("Array", item);
    }

    function toLabel(url) {
        ///<summary>Converts a url to a file label</summary>
        var items = url.split("/"),
             name = items[items.length - 1],
             i    = name.indexOf("?");

        return i !== -1 ? name.substring(0, i) : name;
    }

    // INFO: this look like a "im triggering callbacks all over the place, but only wanna run it one time function" ..should try to make everything work without it if possible
    // INFO: Even better. Look into promises/defered's like jQuery is doing
    function one(callback) {
        ///<summary>Execute a callback only once</summary>
        callback = callback || noop;

        if (callback._done) {
            return;
        }

        callback();
        callback._done = 1;
    }
    //#endregion
    
    function conditional(test, success, failure, callback) {
        ///<summary>
        /// INFO: use cases:
        ///    head.test(condition, null       , "file.NOk" , callback);
        ///    head.test(condition, "fileOk.js", null       , callback);
        ///    head.test(condition, "fileOk.js", "file.NOk" , callback);
        ///    head.test(condition, "fileOk.js", ["file.NOk", "file.NOk"], callback);
        ///    head.test({
        ///               test    : condition,
        ///               success : [{ label1: "file1Ok.js"  }, { label2: "file2Ok.js" }],
        ///               failure : [{ label1: "file1NOk.js" }, { label2: "file2NOk.js" }],
        ///               callback: callback
        ///    );
        ///    head.test({
        ///               test    : condition,
        ///               success : ["file1Ok.js" , "file2Ok.js"],
        ///               failure : ["file1NOk.js", "file2NOk.js"],
        ///               callback: callback
        ///    );
        ///</summary>
        var obj = (typeof test === "object") ? test : {
            test: test,
            success: !!success ? isArray(success) ? success : [success] : false,
            failure: !!failure ? isArray(failure) ? failure : [failure] : false,
            callback: callback || noop
        };

        // Test Passed ?
        var passed = !!obj.test;

        // Do we have a success case
        if (passed && !!obj.success) {
            obj.success.push(obj.callback);
            api.load.apply(null, obj.success);
        }
            // Do we have a fail case
        else if (!passed && !!obj.failure) {
            obj.failure.push(obj.callback);
            api.load.apply(null, obj.failure);
        }
        else {
            callback();
        }

        return api;
    }

    function getAsset(item) {
        ///<summary>
        /// Assets are in the form of
        /// {
        ///     name : label,
        ///     url  : url,
        ///     state: state
        /// }
        ///</summary>
        var asset = {};

        if (typeof item === "object") {
            for (var label in item) {
                if (!!item[label]) {
                    asset = {
                        name: label,
                        url : item[label]
                    };
                }
            }
        }
        else {
            asset = {
                name: toLabel(item),
                url : item
            };
        }

        // is the item already existant
        var existing = assets[asset.name];
        if (existing && existing.url === asset.url) {
            return existing;
        }

        assets[asset.name] = asset;
        return asset;
    }

    function allLoaded(items) {
        items = items || assets;

        for (var name in items) {
            if (items.hasOwnProperty(name) && items[name].state !== LOADED) {
                return false;
            }
        }
        
        return true;
    }

    function onPreload(asset) {
        asset.state = PRELOADED;

        each(asset.onpreload, function (afterPreload) {
            afterPreload.call();
        });
    }

    function preLoad(asset, callback) {
        if (asset.state === undefined) {

            asset.state     = PRELOADING;
            asset.onpreload = [];

            loadAsset({ url: asset.url, type: "cache" }, function () {
                onPreload(asset);
            });
        }
    }

    function apiLoadHack() {
        /// <summary>preload with text/cache hack
        ///
        /// head.load("http://domain.com/file.js","http://domain.com/file.js", callBack)
        /// head.load({ label1: "http://domain.com/file.js" }, { label2: "http://domain.com/file.js" }, callBack)
        /// </summary>

        var args = arguments,
            rest = [].slice.call(args, 1),
            next = rest[0];

        // wait for a while. immediate execution causes some browsers to ignore caching
        if (!isHeadReady) {
            queue.push(function () {
                api.load.apply(null, args);
            });

            return api;
        }

        // multiple arguments
        if (!!next) {
            /* Preload with text/cache hack (not good!)
             * http://blog.getify.com/on-script-loaders/
             * http://www.nczonline.net/blog/2010/12/21/thoughts-on-script-loaders/
             * If caching is not configured correctly on the server, then items could load twice !
             *************************************************************************************/
            each(rest, function (item) {
                // item is not a callback or empty string
                if (!isFunction(item) && !!item) {
                    preLoad(getAsset(item));
                }
            });

            // execute
            load(getAsset(args[0]), isFunction(next) ? next : function () {
                api.load.apply(null, rest);
            });
        }
        else {
            // single item
            load(getAsset(args[0]));
        }

        return api;
    }

    function apiLoadAsync() {
        ///<summary>
        /// simply load and let browser take care of ordering
        ///
        ///    head.load("http://domain.com/file.js","http://domain.com/file.js", callBack)
        ///    head.load({ label1: "http://domain.com/file.js" }, { label2: "http://domain.com/file.js" }, callBack)
        ///</summary>
        var args = arguments,
             callback = args[args.length - 1],
             items = {};

        if (!isFunction(callback)) {
            callback = null;
        }

        each(args, function (item, i) {
            if (item !== callback) {
                item = getAsset(item);
                items[item.name] = item;

                load(item, callback && i === args.length - 2 ? function () {
                    if (allLoaded(items)) {
                        one(callback);
                    }

                } : null);
            }
        });

        return api;
    }

    function load(asset, callback) {
        ///<summary>Used with normal loading logic</summary>
        callback = callback || noop;

        if (asset.state === LOADED) {
            callback();
            return;
        }

        // INFO: why would we trigger a ready event when its not really loaded yet ?
        if (asset.state === LOADING) {
            api.ready(asset.name, callback);
            return;
        }

        if (asset.state === PRELOADING) {
            asset.onpreload.push(function () {
                load(asset, callback);
            });
            return;
        }

        asset.state = LOADING;
        
        loadAsset(asset, function () {
            asset.state = LOADED;
            callback();

            // handlers for this asset
            each(handlers[asset.name], function (fn) {
                one(fn);
            });

            // dom is ready & no assets are queued for loading
            // INFO: shouldn't we be doing the same test above ?
            if (isDomReady && allLoaded()) {
                each(handlers.ALL, function (fn) {
                    one(fn);
                });
            }
        });
    }

    /* Parts inspired from: https://github.com/cujojs/curl
    ******************************************************/
    function loadAsset(asset, callback) {
        callback = callback || noop;
                
        function error(event) {
            event = event || win.event;

            // release event listeners
            ele.onload = ele.onreadystatechange = ele.onerror = null;

            // do callback
            callback();
            
            // need some more detailed error handling here
        }
        
        function process(event) {
            event = event || win.event;
            
            // IE 7/8 (2 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = readystatechange, s.readyState = loaded

            // IE 7/8 (1 event on reload)
            // 1) event.type = readystatechange, s.readyState = complete

            // event.type === 'readystatechange' && /loaded|complete/.test(s.readyState)

            // IE 9 (3 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = readystatechange, s.readyState = loaded
            // 3) event.type = load            , s.readyState = loaded

            // IE 9 (2 events on reload)
            // 1) event.type = readystatechange, s.readyState = complete
            // 2) event.type = load            , s.readyState = complete

            // event.type === 'load'             && /loaded|complete/.test(s.readyState)
            // event.type === 'readystatechange' && /loaded|complete/.test(s.readyState)

            // IE 10 (3 events on 1st load)
            // 1) event.type = readystatechange, s.readyState = loading
            // 2) event.type = load            , s.readyState = complete
            // 3) event.type = readystatechange, s.readyState = loaded

            // IE 10 (3 events on reload)
            // 1) event.type = readystatechange, s.readyState = loaded
            // 2) event.type = load            , s.readyState = complete
            // 3) event.type = readystatechange, s.readyState = complete

            // event.type === 'load'             && /loaded|complete/.test(s.readyState)
            // event.type === 'readystatechange' && /complete/.test(s.readyState)

            // Other Browsers (1 event on 1st load)
            // 1) event.type = load, s.readyState = undefined

            // Other Browsers (1 event on reload)
            // 1) event.type = load, s.readyState = undefined

            // event.type == 'load' && s.readyState = undefined

            // !doc.documentMode is for IE6/7, IE8+ have documentMode
            if (event.type === "load" || (/loaded|complete/.test(ele.readyState) && (!doc.documentMode || doc.documentMode < 9))) {
                // release event listeners
                ele.onload = ele.onreadystatechange = ele.onerror = null;
                
                // do callback
                callback();
            }

            // emulates error on browsers that don't create an exception
            // INFO: timeout not clearing ..why ?
            //asset.timeout = win.setTimeout(function () {
            //    error({ type: "timeout" });
            //}, 3000);
        }

        var ele;
        if (/\.css[^\.]*$/.test(asset.url)) {
            ele      = doc.createElement("link");
            ele.type = "text/" + (asset.type || "css");
            ele.rel  = "stylesheet";
            ele.href = asset.url;
        }
        else {
            ele      = doc.createElement("script");
            ele.type = "text/" + (asset.type || "javascript");
            ele.src  = asset.url;
        }

        ele.onload  = ele.onreadystatechange = process;
        ele.onerror = error;

        /* Good read, but doesn't give much hope !
         * http://blog.getify.com/on-script-loaders/
         * http://www.nczonline.net/blog/2010/12/21/thoughts-on-script-loaders/
         * https://hacks.mozilla.org/2009/06/defer/
         */

        // ASYNC: load in parellel and execute as soon as possible
        ele.async = false;
        // DEFER: load in parallel but maintain execution order
        ele.defer = false;

        // use insertBefore to keep IE from throwing Operation Aborted (thx Bryan Forbes!)
        var head = doc.head || doc.getElementsByTagName("head")[0];
        // but insert at end of head, because otherwise if it is a stylesheet, it will not override values
        head.insertBefore(ele, head.lastChild);
    }

    /* Parts inspired from: https://github.com/jrburke/requirejs
    ************************************************************/
    function init() {
        var items =  doc.getElementsByTagName("script");
        
        //Look for a script with a data-head-init attribute
        for (var i = 0, l = items.length; i < l; i++) {
            var dataMain = items[i].getAttribute("data-headjs-load");
            if (!!dataMain) {
                api.load(dataMain);
                return;
            }
        }
    }
    
    function ready(key, callback) {
        ///<summary>
        /// INFO: use cases:
        ///    head.ready(callBack)
        ///    head.ready(document , callBack)
        ///    head.ready("file.js", callBack);
        ///    head.ready("label"  , callBack);
        ///</summary>

        // DOM ready check: head.ready(document, function() { });
        if (key === doc) {
            if (isDomReady) {
                one(callback);
            }
            else {
                domWaiters.push(callback);
            }

            return api;
        }

        // shift arguments
        if (isFunction(key)) {
            callback = key;
            key      = "ALL";
        }

        // make sure arguments are sane
        if (typeof key !== "string" || !isFunction(callback)) {
            return api;
        }

        // This can also be called when we trigger events based on filenames & labels
        var asset = assets[key];

        // item already loaded --> execute and return
        if (asset && asset.state === LOADED || key === "ALL" && allLoaded() && isDomReady) {
            one(callback);
            return api;
        }

        var arr = handlers[key];
        if (!arr) {
            arr = handlers[key] = [callback];
        }
        else {
            arr.push(callback);
        }

        return api;
    }

    /* Mix of stuff from jQuery & IEContentLoaded
     * http://dev.w3.org/html5/spec/the-end.html#the-end
     ***************************************************/
    function domReady() {
        // Make sure body exists, at least, in case IE gets a little overzealous (jQuery ticket #5443).
        if (!doc.body) {
            // let's not get nasty by setting a timeout too small.. (loop mania guaranteed if assets are queued)
            win.clearTimeout(api.readyTimeout);
            api.readyTimeout = win.setTimeout(domReady, 50);
            return;
        }

        if (!isDomReady) {
            isDomReady = true;
            
            init();
            each(domWaiters, function (fn) {
                one(fn);
            });
        }
    }

    function domContentLoaded() {
        // W3C
        if (doc.addEventListener) {
            doc.removeEventListener("DOMContentLoaded", domContentLoaded, false);
            domReady();
        }

        // IE
        else if (doc.readyState === "complete") {
            // we're here because readyState === "complete" in oldIE
            // which is good enough for us to call the dom ready!
            doc.detachEvent("onreadystatechange", domContentLoaded);
            domReady();
        }
    }

    // Catch cases where ready() is called after the browser event has already occurred.
    // we once tried to use readyState "interactive" here, but it caused issues like the one
    // discovered by ChrisS here: http://bugs.jquery.com/ticket/12282#comment:15
    if (doc.readyState === "complete") {
        domReady();
    }

    // W3C
    else if (doc.addEventListener) {
        doc.addEventListener("DOMContentLoaded", domContentLoaded, false);

        // A fallback to window.onload, that will always work
        win.addEventListener("load", domReady, false);
    }

    // IE
    else {
        // Ensure firing before onload, maybe late but safe also for iframes
        doc.attachEvent("onreadystatechange", domContentLoaded);

        // A fallback to window.onload, that will always work
        win.attachEvent("onload", domReady);

        // If IE and not a frame
        // continually check to see if the document is ready
        var top = false;

        try {
            top = !win.frameElement && doc.documentElement;
        } catch (e) { }

        if (top && top.doScroll) {
            (function doScrollCheck() {
                if (!isDomReady) {
                    try {
                        // Use the trick by Diego Perini
                        // http://javascript.nwbox.com/IEContentLoaded/
                        top.doScroll("left");
                    } catch (error) {
                        // let's not get nasty by setting a timeout too small.. (loop mania guaranteed if assets are queued)
                        win.clearTimeout(api.readyTimeout);
                        api.readyTimeout = win.setTimeout(doScrollCheck, 50);
                        return;
                    }

                    // and execute any waiting functions
                    domReady();
                }
            }());
        }
    }
    //#endregion

    //#region Public Exports
    // INFO: determine which method to use for loading
    api.load  = api.js = isAsync ? apiLoadAsync : apiLoadHack;
    api.test  = conditional;
    api.ready = ready;
    //#endregion

    //#region INIT
    // perform this when DOM is ready
    api.ready(doc, function () {

        if (allLoaded()) {
            each(handlers.ALL, function (callback) {
                one(callback);
            });
        }

        if (api.feature) {
            api.feature("domloaded", true);
        }
    });

    /*
        We wait for 300 ms before asset loading starts. for some reason this is needed
        to make sure assets are cached. Not sure why this happens yet. A case study:

        https://github.com/headjs/headjs/issues/closed#issue/83
    */
    setTimeout(function () {
        isHeadReady = true;
        each(queue, function (fn) {
            fn();
        });

    }, 300);
    //#endregion
}(window));